### PR TITLE
Psycopg 3 support

### LINKED ISSUE #499: Add Psycopg 3 support
##### SUMMARY
- Related PR https://github.com/ansible-collections/community.postgresql/pull/220 that can be used when implementing actual support (the obstacle was fixed on the driver's stuff)
- Another big related issue to solve it adding testing against psycopg 3, i.e. to have testing against 2 and 3

### PR DESCRIPTION
**Pull request by elpavel**
##### SUMMARY
Added support for Psycopg 3

Fixes #499

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
All modules

##### ADDITIONAL INFORMATION
Work in progress, any feedback welcome

### CODE REVIEW

**Review by hunleyd — CHANGES_REQUESTED** [review]
the failed tests are throwing `has no attribute 'psycopg2'` which I assume means you're not done @elpavel ?

**Comment by elpavel** [comment]
Hi @hunleyd this is just a start. I have updated the common postgres.py first, to see what breaks. Now I need to fix the units tests for it. Then I plan to update the modules needed for test db build - db, user, set. And then all the remaining modules. 
The unit tests should be easy to fix, after that only the two Ubuntu 22.04 builds should keep failing until all modules are updated with Psycopg 3.

**Comment by Andersson007** [comment]
FYI: I marked the PR as non-draft to run the tests. I think `WIP` in description is a good indicator that it's WIP.
@elpavel just ping us here and on Matrix when it's ready for review
Great undertaking, thanks!

**Comment by elpavel** [comment]
Adding the Psycopg 3 support will require changes in all modules. The changes could be grouped into these categories:
* handling of connection.server_version
```
-    if self.cursor.connection.server_version >= 100000:
+    if get_server_version(self.cursor.connection) >= 100000:
```
* different parameters for connection.cursor ... so far I am solving by defining the params in postgres.py, this also removes the need for most modules to import anything from psycopg
```
-    cursor = db_connection.cursor(cursor_factory=DictCursor)
+    cursor = db_connection.cursor(**pg_cursor_args)
```
* changes in comments, using psycopg instead of psycopg2
* psycopg 3 result set is a dict, so we need to use column names as keys
```
-    if not res[0][0]:
+    if not res[0]["exists"]:
```
* acl type columns are handled differently, we can cast them to string to preserver psycopg2 logic
```
-    query = """SELECT nspacl FROM pg_catalog.pg_namespace
+    query = """SELECT nspacl::varchar FROM pg_catalog.pg_namespace
```
* error codes, not sure yet what is the equivalent of pgcode from psycopg2
* the actual psycopg3 import
* test coverage

The above will cover majority of changes. I think we could introduce these changes incrementally, ahead of adding the actual psycopg3 import, as in #518.

File: plugins/module_utils/postgres.py [hunk 1]
```diff
@@ -16,19 +16,40 @@
...
+try:
+    import psycopg
+    from psycopg import ClientCursor
+    from psycopg.rows import dict_row
+    # We need Psycopg 3 to be at least 3.1.0 because we need Client-side-binding cursors
+    # When a Linux distribution provides both Psycopg2 and Psycopg 3.0 we will use Psycopg2
+    PSYCOPG_VERSION = LooseVersion(psycopg.__version__)
+    if PSYCOPG_VERSION < LooseVersion("3.1"):
+        raise ImportError
+    HAS_PSYCOPG = True
```
[the remark this answers is not in the document]
Comment (hunleyd) [reply plugins/module_utils/postgres.py:38]: or is HAS_PSYCOPG supposed to indicate that we have _some_ version of the lib, regardless of which version?

[the remark this answers is not in the document]
Comment (elpavel) [reply plugins/module_utils/postgres.py:38]: Correct, HAS_PSYCOPG indicates we have some version.
I think HAS_PSYCOPG together with PSYCOPG_VERSION will cover all options.

**Comment by Andersson007** [comment]
great to see the progress here!

File: .azure-pipelines/azure-pipelines.yml [hunk 2]
```diff
@@ -115,8 +115,8 @@ stages:
           targets:
             - name: CentOS 7
               test: centos7
-            - name: Fedora 37
-              test: fedora37
+            - name: Fedora 38
+              test: fedora38
```
Comment (Andersson007) [inline .azure-pipelines/azure-pipelines.yml:119]:
```suggestion
            - name: Fedora 38
              test: fedora38
```
@elpavel should we maybe add this in a separate PR and merge right now? (and the associated line you added to README)
If you agree, feel free to raise a PR

Comment (Andersson007) [reply .azure-pipelines/azure-pipelines.yml:119]: i'm not sure if it will work though:)

Comment (elpavel) [reply .azure-pipelines/azure-pipelines.yml:119]: It should work, I will raise the PR for adding Fedora 38 as a test target.

Fedora 38 has packages for both Psycopg 2 and 3, our normal testing setup will use Psycopg 2 and all should work as usual. Psycopg 3 in Fedora 38 is 3.1 which is good for the new code testing as 3.0 doesn't work.

BTW Fedora 36 and 37 also have packages available for Psycopg 3, but the version is 3.0.

Comment (Andersson007) [reply .azure-pipelines/azure-pipelines.yml:119]: cool, thanks!

Comment (elpavel) [reply .azure-pipelines/azure-pipelines.yml:119]: Created #527 to cover this

**Comment by Andersson007** [comment]
@elpavel i don't know if you aware of the fact that sometimes when you commit things in other PRs that can affect say CI in another PR, you can [rebase](https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html) the latter from time to time. It's 100% relevant for collections with GHA but IIRC we noticed something similar here.
Just a fyi

**Comment by Andersson007** [comment]
> > @elpavel i don't know if you aware of the fact that sometimes when you commit things in other PRs that can affect say CI in another PR, you can [rebase](https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html) the latter from time to time. It's 100% relevant for collections with GHA but IIRC we noticed something similar here. Just a fyi
> 
> I keep merging changes from main into this one from time to time, is that what you mean, or something else?
> 
> I see now, tried to rebase this branch, looks ok.

yep, exactly, some contributors don't know about it, so mentioned it just in case

**Comment by Andersson007** [comment]
@elpavel great epic work! Could you please remove `WIP`? Anyway, i'll setup a reminder for Monday to review the PR. Thanks!

**Review by hunleyd — APPROVED** [review]
looks sane to me. nicely done!

File: tests/integration/targets/setup_postgresql_db/vars/Fedora-36-py3.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# We install both psycopg2 and psycopg3.
+# As psycopg3 is only 3.0, the modules should use psycopg2.
```
[the remark this answers is not in the document]
Comment (elpavel) [reply tests/integration/targets/setup_postgresql_db/vars/Fedora-36-py3.yml:2]: The intention here is to install both rpm packages, versions 2 and 3, as if someone installs all rpms. It is a test for the code not to use 3.0, and to use psycopg2 instead.

File: plugins/module_utils/postgres.py [hunk 4]
```diff
@@ -106,16 +121,25 @@ def connect_to_db(module, conn_params, autocommit=False, fail_on_conn=True):
     db_connection = None
     conn_err = None
     try:
-        db_connection = psycopg2.connect(**conn_params)
-        if autocommit:
-            if PSYCOPG_VERSION >= LooseVersion("2.4.2"):
-                db_connection.set_session(autocommit=True)
-            else:
-                db_connection.set_isolation_level(psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT)
+        if PSYCOPG_VERSION >= LooseVersion("3.0"):
```
[the remark this answers is not in the document]
Comment (elpavel) [reply plugins/module_utils/postgres.py:124]: It could be anything from "3", "3.0", "3.1", the meaning is "not psycopg2".

**Review by Andersson007 — APPROVED** [review]

**Pull request merged**